File: day17/Day17.scala
import scala.io.Source.stdin

@main def main(idx: Int) =
  val Seq(xmin, xmax, ymin, ymax) = raw"-?\d+".r.findAllIn(stdin.getLines.next)
                                                .map{_.toInt}.toSeq
  var maxY, hitCount = 0
  for { sxv <- 0 to (xmax + 1); syv <- ymin to 300 } do
    var (x, y, xv, yv, curMaxY, hit) = (0, 0, sxv, syv, 0, false)
    while !(y < ymin || x > xmax || (xv == 0 && x < xmin)) do
      x += xv; y += yv
      xv -= xv.sign; yv -= 1
      curMaxY = curMaxY.max(y)
      hit |= xmin <= x && x <= xmax && ymin <= y && y <= ymax
    if hit then
      maxY = maxY.max(curMaxY)
      hitCount += 1

  println{ if (idx == 1) maxY else hitCount }

File: day17/Day17.rs
use std::env;
use std::io;

fn main() {
    let desc = io::stdin().lines().next().unwrap().unwrap();
    let parts: Vec<i32> = desc
        .split(|c| matches!(c, '=' | '.' | ',' | ' '))
        .filter_map(|s| s.parse().ok())
        .collect();

    let guess = 300;

    let mut total_max_y = 0;
    let mut hit_count = 0;

    if let [xmin, xmax, ymin, ymax] = parts[..] {
        for xv in 0..=xmax + 1 {
            for yv in ymin - 1..=guess {
                let mut xv = xv;
                let mut yv = yv;
                let mut x = 0;
                let mut y = 0;

                let mut max_y = 0;
                let mut hit = false;

                loop {
                    x += xv;
                    y += yv;

                    xv -= xv.signum();
                    yv -= 1;

                    max_y = max_y.max(y);

                    if xmin <= x && x <= xmax && ymin <= y && y <= ymax {
                        hit = true;
                    }

                    if y < ymin || x > xmax || (xv == 0 && x < xmin) {
                        break;
                    }
                }

                if hit {
                    total_max_y = total_max_y.max(max_y);
                    hit_count += 1;
                }
            }
        }
    }

    if env::args().nth(1).unwrap() == "1" {
        println!("{}", total_max_y);
    } else {
        println!("{}", hit_count);
    }
}
